proxy/src/control/discovery.rs

```diff
@@ -391,79 +398,152 @@ impl<T: HttpService<RequestBody = BoxBody, ResponseBody = RecvBody>> Destination
-        trace!("remove {:?} for {:?}", addr, authority_for_logging);
+    fn on_change(txs: &mut Vec<mpsc::UnboundedSender<Update>>,
+                 authority_for_logging: &FullyQualifiedAuthority,
+                 addr: SocketAddr,
+                 change: CacheChange) {
+        let (update_str, update_constructor): (&'static str, fn(SocketAddr) -> Update) =
+            match change {
+                CacheChange::Insertion => ("insert", Update::Insert),
+                CacheChange::Removal => ("remove", Update::Remove),
+            };
+        trace!("{} {:?} for {:?}", update_str, addr, authority_for_logging);
         // retain is used to drop any senders that are dead
-        self.txs.retain(|tx| {
-            tx.unbounded_send(Update::Remove(addr)).is_ok()
+        txs.retain(|tx| {
+            tx.unbounded_send(update_constructor(addr)).is_ok()
         });
     }
 }
 
+// ===== impl Cache =====
+
+enum CacheChange {
+    Insertion,
+    Removal,
+}
+
+impl<T> Cache<T> where T: Clone + Copy + Eq + std::hash::Hash {
+    fn new() -> Self {
+        Cache {
+            values: HashSet::new(),
+            reset_on_next_modification: true,
+        }
+    }
+
+    fn extend<I, F>(&mut self, iter: I, on_change: &mut F)
+        where I: Iterator<Item = T>,
+              F: FnMut(T, CacheChange),
+    {
+        fn extend_inner<T, I, F>(values: &mut HashSet<T>, iter: I, on_change: &mut F)
+            where T: Copy + Eq + std::hash::Hash, I: Iterator<Item = T>, F: FnMut(T, CacheChange)
+        {
+            for value in iter {
+                if values.insert(value) {
+                    on_change(value, CacheChange::Insertion);
+                }
+            }
+        }
+
+        if !self.reset_on_next_modification {
+            extend_inner(&mut self.values, iter, on_change);
+        } else {
+            let to_insert = iter.collect::<HashSet<T>>();
+            extend_inner(&mut self.values, to_insert.iter().map(|value| *value), on_change);
+            self.retain(&to_insert, on_change);
+        }
+        self.reset_on_next_modification = false;
+    }
+
+    fn remove<I, F>(&mut self, iter: I, on_change: &mut F)
+        where I: Iterator<Item = T>,
+              F: FnMut(T, CacheChange)
+    {
+        if !self.reset_on_next_modification {
+            for value in iter {
+                if self.values.remove(&value) {
+                    on_change(value, CacheChange::Removal);
+                }
+            }
+        } else {
+            self.clear(on_change);
+        }
+        self.reset_on_next_modification = false;
+    }
+
+    fn clear<F>(&mut self, on_change: &mut F) where F: FnMut(T, CacheChange) {
+        self.retain(&HashSet::new(), on_change)
+    }
+
+    fn retain<F>(&mut self, to_retain: &HashSet<T>, mut on_change: F)
+        where F: FnMut(T, CacheChange)
+    {
+        for value in self.values.clone().difference(to_retain) {
```

Comment:
fwiw, i think we could likely avoid this clone if we switched to IndexSet (instead of HashSet).  not a blocker

Reply:
Hm, is it faster to clone the full hashset and call `difference`, instead of using `HashSet::retain` directly?

```rust
self.values.retain(|val| {
    to_retain.contains(val)
});
```

Reply:
We have to call `on_change` for each removal, so we can't use `retain()` unless I'm misunderstanding something.

Reply:
So, couldn't be this?

```rust
self.values.retain(|val| {
    if !to_retain.contains(val) {
        on_change(*val, CacheChange::Removal);
        false
    } else {
        true
    }
});
```

Reply:
OK, gotcha. Thanks!